pkg/api/v1/control-plane/cache/prioritylist.go

```diff
@@ -0,0 +1,165 @@
+package cache
+
+import (
+	"fmt"
+	"sort"
+)
+
+// The default Priority is the last index. The last index is not required in the prioirty list, thus any elements added
+// to the struct will be added to the default priority (last).
+const defaultPrioirty = -1
+
+type PriorityValue interface {
+	GetPriority() string
+}
+
+// PrioritySortedStruct sorts elements by prioirity. Priority lists are unordered.
+// Needs a priority list and a get priority function.
+// If a priority does not exist for an added element, it is added to the lowest priority.
+type PrioritySortedStruct struct {
+	// prioritySets is the set lists for priorities, where P is the type used for priority.
+	prioritySets map[int][]string
+	// priorityMap maps the value to the priority.
+	priorityMap map[string]int
+	// elements are the map of elements structured by their priority.
+	elements []map[int]ResponseWatch
+	// numberOfPriorities is the number of priorities in the prioritySets
+	numberOfPriorities int
+	// nextUniqueElementIndex is the next element index to be inserted.
+	nextUniqueElementIndex int
+}
+
+// PriorityIndex is the priority and index used to locate items.
+type PriorityIndex struct {
+	Priority int
+	Index    int
+}
+
+func NewPrioritySortedStruct(prioritySets map[int][]string) *PrioritySortedStruct {
+	p := PrioritySortedStruct{
+		prioritySets: prioritySets,
+	}
+	p.Init()
+	return &p
+}
+
+func (p *PrioritySortedStruct) Init() {
+	p.numberOfPriorities = len(p.prioritySets)
+	// need to ensure that the prioriries are in order and there are no missing or skipped Priorities
+	priorities := make([]int, 0)
+	for priority, _ := range p.prioritySets {
+		priorities = append(priorities, priority)
+	}
+	sort.Ints(priorities)
+	currentP := 0
+	for _, p := range priorities {
+		if currentP == p {
+			currentP++
+		} else {
+			panic(fmt.Sprintf("Priorities are not set correct, you are missing priority %d", currentP))
+		}
+	}
+	p.elements = make([]map[int]ResponseWatch, p.numberOfPriorities+1)
+	if p.prioritySets != nil && p.numberOfPriorities > 0 {
```

Comment:
Why are we doing all this set up ahead of time? Cant we just check if the given prio exists at insert time and add it then?
That being said not against this entirely

Reply:
Not sure why it matters